[LAA] Keep pointer checks on partial analysis

john-brawn-arm opened the pull request:
Currently if there's any memory access that AccessAnalysis couldn't analyze then all of the runtime pointer check results are discarded. This patch changes that so they're not, and we generate the runtime check information for those pointers that we could analyze, as transformations may still be able to make use of the partial information.

Of the transformations that use LoopAccessAnalysis, only LoopVersioningLICM changes behaviour as a result of this change. This is because the others either:
 * Check canVectorizeMemory, which will return false when we have partial pointer information as analyzeLoop() will return false.
 * Examine the dependencies returned by getDepChecker(), which will be empty as we exit analyzeLoop if we have partial pointer information before calling areDepsSafe(), which is what fills in the dependency information.

john-brawn-arm committed:
[LoopAccessAnalysis] Keep pointer checks on partial analysis

Currently if there's any memory access that AccessAnalysis couldn't
analyze then all of the runtime pointer check results are discarded.
This patch changes that so they're not, and we generate the runtime
check information for those pointers that we could analyze, as
transformations may still be able to make use of the partial
information.

Of the transformations that use LoopAccessAnalysis, only
LoopVersioningLICM changes behaviour as a result of this change. This
is because the others either:
 * Check canVectorizeMemory, which will return false when we have
   partial pointer information as analyzeLoop() will return false.
 * Examine the dependencies returned by getDepChecker(), which will be
   empty as we exit analyzeLoop if we have partial pointer information
   before calling areDepsSafe(), which is what fills in the dependency
   information.

artagnon commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1313,7 +1316,7 @@ bool AccessAnalysis::canCheckPtrAtRT(
     }
   }
 
-  if (MayNeedRTCheck && CanDoRT)
+  if (MayNeedRTCheck)
```
Not sure I understand this change: why don't we just generate RT checks if we may need RT checks? Is it an unrelated bugfix?

john-brawn-arm replied:
CanDoRT will be false when createCheckForAccess returned false for some memory accesses meaning the current behaviour is to not generate any checks if there was any access that createCheckForAccess was false for. Removing the && CanDoRT means that we will generate checks for those that createCheckForAccess did return true for.

artagnon reviewed (commented):
I think this is a sound change, and the test updates to LoopVersioningLICM look good!

john-brawn-arm commented on llvm/test/Analysis/LoopAccessAnalysis/forked-pointers.ll:
```diff
@@ -633,7 +633,18 @@ define dso_local void @forked_ptrs_same_base_different_offset(ptr nocapture read
 ; CHECK-NEXT:      Report: cannot identify array bounds
```
This piece of text is what goes in the -Rpass-analysis=loop-vectorize output, so here it's
```
remarks.c:8:12: remark: loop not vectorized: cannot identify array bounds [-Rpass-analysis=loop-vectorize]
    8 |     B[i] = A[offset];
      |            ^
```
It's intended to mean "cannot identify array bounds of this specific memory access".

artagnon replied:
Oh. Hm, then is there something we can add to the output to indicate that runtime checks were not generated for some pointers? Otherwise, the output seems a bit misleading.

john-brawn-arm committed:
Add extra line to LoopAccessInfo::print when we generate some but not all checks

artagnon commented on llvm/test/Analysis/LoopAccessAnalysis/pointer-phis.ll:
```diff
@@ -287,6 +287,7 @@ define i32 @store_with_pointer_phi_incoming_phi(ptr %A, ptr %B, ptr %C, i1 %c.0,
 ; CHECK-NEXT:          (Low: %A High: (256000 + %A))
 ; CHECK-NEXT:            Member: {%A,+,8}<nuw><%loop.header>
 ; CHECK-NEXT:            Member: %A
+; CHECK-NEXT:      Generated run-time checks are incomplete
```
Are they incomplete in this case?

artagnon commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -3002,6 +3001,8 @@ void LoopAccessInfo::print(raw_ostream &OS, unsigned Depth) const {
 
   // List the pair of accesses need run-time checks to prove independence.
   PtrRtChecking->print(OS, Depth);
+  if (PtrRtChecking->Need && !CanVecMem)
```
I think this check is a bit too lax, and generates spurious messages. Can you check CheckingGroups against DepCands instead?

artagnon reviewed (commented):
This should be good to go, modulo the spurious messages. I checked LVerLICM, and it seems to do the right thing:

```cpp
    // Don't allow stores that we don't have runtime checks for, as we won't be
    // able to mark them noalias meaning they would prevent any code motion.
    auto &Pointers = LAI->getRuntimePointerChecking()->Pointers;
    if (!any_of(Pointers, [&](auto &P) { return P.PointerValue == Ptr; })) {
      LLVM_DEBUG(dbgs() << "    Found a store without a runtime check.\n");
      return false;
    }
```

fhahn reviewed (commented):
Would it be possible to make this behavior opt-in? Otherwise we unnecessarily increase compile-time for all other users that do not need this, just to benefit LoopVerLICM which isn't on by default.

Expanding low/end for groups can be costly in some cases.

john-brawn-arm commented:
> Would it be possible to make this behavior opt-in? Otherwise we unnecessarily increase compile-time for all other users that do not need this, just to benefit LoopVerLICM which isn't on by default.
> 
> Expanding low/end for groups can be costly in some cases.

Doing a simple test with the following:
{code}
void fn(char *p, char *q, char **r, int n) {
#define LOOP \
  for (int i = 0; i < n; i++) { \
    p[i] += q[i] + r[0][i]; \
    p[i] += q[i+1] + r[1][i]; \
    p[i] += q[i+2] + r[2][i]; \
    p[i] += q[i+3] + r[3][i]; \
    p[i] += q[i+4] + r[4][i]; \
    p[i] += q[i+5] + r[5][i]; \
  }

#define DO10 LOOP LOOP LOOP LOOP LOOP LOOP LOOP LOOP LOOP LOOP
#define DO100 DO10 DO10 DO10 DO10 DO10 DO10 DO10 DO10 DO10 DO10
#define DO1000 DO100 DO100 DO100 DO100 DO100 DO100 DO100 DO100 DO100 DO100

  DO1000
}
{code}
and doing:

- Compile with clang -O2 -emit-llvm
- Run ``opt  -passes="print<access-info>" -disable-output --time-trace`` 20 times with and without this change
- Get the duration of LoopAccessAnalysis (which is in microseconds) from the time trace file

I get:
Without: mean 1307, median 1347, standard deviation 119
With: mean 1301, median 1335, standard deviation 116
So it looks like any increase in compile-time is negligible and well within the measurement noise, for this test case at least.

john-brawn-arm committed:
Merge branch 'main' into loop_access_analysis_partial_result

artagnon commented:
See 234cc40a ([LAA] Limit no-overlap check to at least one loop-invariant accesses.): @fhahn has good reason to believe that compile-time will blow up. From the comment:

```cpp
  // Check if we can prove that Sink only accesses memory after Src's end or
  // vice versa. At the moment this is limited to cases where either source or
  // sink are loop invariant to avoid compile-time increases. This is not
  // required for correctness.
```

john-brawn-arm committed:
Add HasCompletePtrRtChecking

This records whether canCheckPtrAtRT succeeded, and is used to decide
if we print the "checks are incomplete" message.

john-brawn-arm committed:
Merge branch 'main' into loop_access_analysis_partial_result

john-brawn-arm committed:
Add an option to control whether partial results are allowed

john-brawn-arm commented:
LLVM compile-time tracker results: https://llvm-compile-time-tracker.com/compare.php?from=a7ede51b556f40163db9e3cc67c98c27ba2364d8&to=a4dd603c004278e86f80cf6d62597535f0fee126&stat=instructions%3Au
There's a couple of benchmarks (mafft and lencod) that are showing as red, but the largest regression is 0.06% which doesn't seem like a lot. Is it still large enough to need to do something about it?

artagnon commented:
> but the largest regression is 0.06% which doesn't seem like a lot. Is it still large enough to need to do something about it?

Regressions of 0.01-0.02% are noise (the compile time tracker doesn't color them), 0.03% is a noticeable regression (the compile time tracker colors these in light red), and 0.05%+ is serious (compile time tracker colors these in deep red).

john-brawn-arm committed:
Merge branch 'main' into HEAD

artagnon reviewed (approved):
LGTM, thanks! A few minor suggestions follow.

john-brawn-arm committed:
Adjust based on review comments.

fhahn commented on llvm/include/llvm/Analysis/LoopAccessAnalysis.h:
```diff
@@ -784,20 +792,28 @@ class LoopAccessInfo {
 
   Loop *TheLoop;
 
+  bool AllowPartial;
```
Needs some documentation. As AllowPartial is the default, it might be better to have the flag indicate that all runtime that can be generated will be collected?

fhahn replied:
What about the naming?

john-brawn-arm replied:
The comment was added in https://github.com/llvm/llvm-project/pull/139719/commits/2b02a5bf6671d1feef49c49c947b7083e98e393b. It looks like github isn't showing the comment on the "conversations" tab but does on the "files changed" tab.

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2599,9 +2605,9 @@ bool LoopAccessInfo::analyzeLoop(AAResults *AA, const LoopInfo *LI,
-  bool CanDoRTIfNeeded = Accesses.canCheckPtrAtRT(
-      *PtrRtChecking, TheLoop, SymbolicStrides, UncomputablePtr);
-  if (!CanDoRTIfNeeded) {
+  HasCompletePtrRtChecking = Accesses.canCheckPtrAtRT(
```
Why does this need to be a member variable?  Can we keep the name which still seems accurate?

john-brawn-arm replied:
This is used in LoopAccessInfo::print to decide whether to print the "Generated run-time checks are incomplete" message.

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2978,12 +2988,15 @@ void LoopAccessInfo::print(raw_ostream &OS, unsigned Depth) const {
-  if (Inserted)
-    It->second =
-        std::make_unique<LoopAccessInfo>(&L, &SE, TTI, TLI, &AA, &DT, &LI);
+  // We need to create the LoopAccessInfo if either we don't already have one,
+  // or if it was created with a different value of AllowPartial.
```
Could we keep the original one if it has AllowPartial = true and AllowPartial = false?

It might be clearer if the flag is renamed to indicate that all runtime checks are collected, even if not valid.

john-brawn-arm replied:
I had thought about that, but I think having consistent output is more useful than any minor efficiency gain from not having to recompute the LoopAccessInfo.

john-brawn-arm replied:
Yes, though potentially there could be some transform pass added in the future that uses LoopAccessAnalysis without setting AllowPartial but which behaves differently when there's existing results results with it set.

fhahn commented on llvm/test/Analysis/LoopAccessAnalysis/allow-partial.ll:
```diff
@@ -0,0 +1,99 @@
+; RUN: opt -disable-output -passes='print<access-info><allow-partial>,print<access-info>' %s 2>&1 | FileCheck %s --check-prefixes=ALLOW-BEFORE
```
I don't think we have other print passes that take arguments already, but would it be more consistent to pare this as `print<access-info<allow-partial>>` to indicate this is an argument to access-info instead of `print<>`?

john-brawn-arm committed:
Add an extra comment
